src/Build/Definition/Project.cs

```diff
@@ -2782,17 +2782,18 @@ private List<ProvenanceResult> GetItemProvenance(string itemToMatch, IEnumerable
                     return new List<ProvenanceResult>();
                 }
 
-                return
-                    projectItemElements
+                return projectItemElements
                     .AsParallel()
-                    .AsOrdered()
-                    .Select(i => ComputeProvenanceResult(itemToMatch, i))
+                    .Select((item, index) => ComputeProvenanceResult(itemToMatch, item, index))
+                    .AsSequential()
+                    .OrderBy(pair => pair.Index)
+                    .Select(pair => pair.Result)
                     .Where(r => r != null)
```

Comment:
I don't know if it would matter for this specific case, but it seems like you could do more in parallel and delay the re-sequentialization:

```suggestion
                    .Where(pair => pair.Result is not null)
                    .AsSequential()
                    .OrderBy(pair => pair.Index)
                    .Select(pair => pair.Result)
```

Reply:
Right, and moving filtration before ordering decreases number of elements to be sorted.